auth/password_authenticator.cc

```diff
@@ -209,33 +209,34 @@ future<> password_authenticator::start() {
 
         _stopped = do_after_system_ready(_as, [this] {
             return async([this] {
-                if (legacy_mode(_qp)) {
-                    if (!_superuser_created_promise.available()) {
+                auto ready = [this] () {
+                if (!_superuser_created_promise.available()) {
                         _superuser_created_promise.set_value();
                     }
+                };
+                if (legacy_mode(_qp)) {
                     _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as).get();
 
                     if (any_nondefault_role_row_satisfies(_qp, &has_salted_hash, _superuser).get()) {
                         if (legacy_metadata_exists()) {
                             plogger.warn("Ignoring legacy authentication metadata since nondefault data already exist.");
                         }
-
+                        ready();
                         return;
                     }
 
                     if (legacy_metadata_exists()) {
                         migrate_legacy_metadata().get();
+                        ready();
                         return;
                     }
                     legacy_create_default_if_missing().get();
                 }
                 utils::get_local_injector().inject("password_authenticator_start_pause", utils::wait_for_message(5min)).get();
                 if (!legacy_mode(_qp)) {
                     maybe_create_default_password_with_retries().get();
-                    if (!_superuser_created_promise.available()) {
-                        _superuser_created_promise.set_value();
-                    }
                 }
+                ready();
```

Comment:
Nit: I've got a suggestion on how to do it in a cleaner way: add a continuation to the future returned by `async` where you conditionally set the `_superuser_created_promise`. This way, it will be handled in one place and nobody will have to remember to add further `ready` calls (unlikely, but also makes it easier to read the code, at least in my opinion).